src/utility/svd.jl

```diff
@@ -0,0 +1,207 @@
+)
+    #early return
+    if isempty(blocksectors(t))
+        truncerr = zero(real(scalartype(t)))
+        return _empty_svdtensors(t)..., truncerr
+    end
+
+    S = spacetype(t)
+    Udata, Σdata, Vdata, dims = _compute_svddata!(t, alg)
+    if !isa(trunc, TensorKit.NoTruncation)
+        Σdata, truncerr = _truncate!(Σdata, trunc, p)
+        Udata, Σdata, Vdata, dims = _implement_svdtruncation!(t, Udata, Σdata, Vdata, dims)
+        W = S(dims)
+    else
+        truncerr = abs(zero(scalartype(t)))
+        W = S(dims)
+        if length(domain(t)) == 1 && domain(t)[1] ≅ W
+            W = domain(t)[1]
+        elseif length(codomain(t)) == 1 && codomain(t)[1] ≅ W
+            W = codomain(t)[1]
+        end
+    end
+    return _create_svdtensors(t, Udata, Σdata, Vdata, W)..., truncerr
+end
+
+# Wrapper struct around TensorKit's SVD algorithms
+@kwdef struct FullSVD
+    alg::Union{<:TensorKit.SVD,<:TensorKit.SDD} = TensorKit.SVD()
+    lorentz_broad::Float64 = 0.0
+end
+
+function svdwrap(t::AbstractTensorMap, alg::FullSVD; trunc=notrunc(), kwargs...)
+    # TODO: Replace _tensorkit_svd! with just tsvd eventually to use the full TensorKit machinery
+    return _tensorkit_svd!(copy(t); trunc, alg.alg)
+end
+
+function ChainRulesCore.rrule(
+    ::typeof(svdwrap), t::AbstractTensorMap, alg::FullSVD; trunc=notrunc(), kwargs...
+)
+    tsvd_return, tsvd!_pullback = ChainRulesCore.rrule(tsvd!, t; alg=TensorKit.SVD(), trunc)
+    function svdwrap_fullsvd_pullback(Δ)
+        return tsvd!_pullback(Δ)..., NoTangent()
+    end
+    return tsvd_return, svdwrap_fullsvd_pullback
+end
+
+# Wrapper around Krylov Kit's GKL iterative SVD solver
+@kwdef struct IterSVD
+    alg::KrylovKit.GKL = KrylovKit.GKL(; tol=1e-14, krylovdim=25)
+    howmany::Int = 20
+    lorentz_broad::Float64 = 0.0
+    alg_rrule::Union{GMRES,BiCGStab,Arnoldi} = GMRES(; tol=1e-14)
+end
+
+function svdwrap(t::AbstractTensorMap, alg::IterSVD; trunc=notrunc(), kwargs...)
+    U, S, V, = _tensorkit_svd!(copy(t); trunc, alg)  # TODO: Also replace this with tsvd eventually
+    ϵ = norm(t - U * S * V)  # Compute truncation error separately
+    return U, S, V, ϵ
+end
+
+# Compute SVD data block-wise using KrylovKit algorithm
+function TensorKit._compute_svddata!(t::TensorMap, alg::IterSVD)
+    InnerProductStyle(t) === EuclideanProduct() || throw_invalid_innerproduct(:tsvd!)
+    I = sectortype(t)
+    A = storagetype(t)
+    Udata = SectorDict{I,A}()
+    Vdata = SectorDict{I,A}()
+    dims = SectorDict{I,Int}()
+    local Σdata
+    for (c, b) in blocks(t)
+        x₀ = randn(eltype(b), size(b, 1))
+        Σ, lvecs, rvecs, info = svdsolve(b, x₀, alg.howmany, :LR, alg.alg)
+        if info.converged < alg.howmany  # Fall back to dense SVD if not properly converged
+            U, Σ, V = TensorKit.MatrixAlgebra.svd!(b, TensorKit.SVD())
+            Udata[c] = U
+            Vdata[c] = V
+        else
+            Udata[c] = stack(lvecs)
+            Vdata[c] = stack(rvecs)'
```

Comment:
I think this implicitly converts the lazy adjoint to a regular matrix, I'd maybe like to make this a bit more explicit